language/diem-framework/modules/NetworkIdentity.move

```diff
@@ -0,0 +1,143 @@
+/// Module managing Diemnet NetworkIdentity
+module DiemFramework::NetworkIdentity {
+    use DiemFramework::DiemTimestamp;
+    use Std::Errors;
+    use Std::Event::{Self, EventHandle};
+    use Std::Signer;
+    use Std::Vector;
+
+    /// Holder for all `NetworkIdentity` in an account
+    struct NetworkIdentity has key {
+        identities: vector<vector<u8>>,
+        /// Event handle for `identities` rotation events
+        identity_change_events: EventHandle<NetworkIdentityChangeNotification>
+    }
+
+    /// Message sent when there are updates to the `NetworkIdentity`.
+    struct NetworkIdentityChangeNotification has drop, store {
+        /// The new identities
+        identities: vector<vector<u8>>,
+        /// The time at which the `identities` was rotated
+        time_rotated_seconds: u64,
+    }
+
+    const MAX_ADDR_IDENTITIES: u64 = 100;
+
+    // Error Codes
+    /// Limit exceeded on number of identities for an address
+    const ENET_ID_LIMIT_EXCEEDED: u64 = 0;
+
+    // =================================================================
+    // Module Specification
+
+    spec module {} // Switch to module documentation context
+
+    /// Initialize `NetworkIdentity` with an empty list
+    fun initialize_network_identity(account: &signer) {
+        let identities = Vector::empty<vector<u8>>();
+        let identity_change_events = Event::new_event_handle<NetworkIdentityChangeNotification>(account);
+        move_to(account, NetworkIdentity { identities, identity_change_events });
+    }
+    spec initialize_network_identity {
+        pragma opaque;
+        let account_addr = Signer::spec_address_of(account);
+        ensures exists<NetworkIdentity>(account_addr);
+        modifies global<NetworkIdentity>(account_addr);
+    }
+
+    /// Return the underlying `NetworkIdentity` bytes
+    public fun get(account_addr: address): vector<vector<u8>> acquires NetworkIdentity {
+        *&borrow_global<NetworkIdentity>(account_addr).identities
+    }
+
+    spec get {
+    }
+
+    /// Update and create if not exist `NetworkIdentity`
+    public fun add_identities(account: &signer, to_add: vector<vector<u8>>) acquires NetworkIdentity {
+        if (!exists<NetworkIdentity>(Signer::address_of(account))) {
+            initialize_network_identity(account);
+        };
+        let identity = borrow_global_mut<NetworkIdentity>(Signer::address_of(account));
+        let identities = &mut identity.identities;
+        let i = 0;
+        let has_change = false;
+        while (i < Vector::length(&to_add)) {
+           has_change = has_change || add_identity(identities, *Vector::borrow(&to_add, i));
+           i = i + 1;
+        };
+
+        assert(Vector::length(identities) <= MAX_ADDR_IDENTITIES, Errors::limit_exceeded(ENET_ID_LIMIT_EXCEEDED));
```

Comment:
Could you move this assertion above where the work is done? I'm worried that an attacker could provide a very long list of `to_add` identities, and it will only be caught after performing all of the work. i.e., add 
```rust
assert(Vector::length(identities) + Vector::length(&to_add) <= MAX_ADDR_IDENTITIES, ...);
```
before the loop.